### ISSUE TITLE
Editorial: fos record descriptions within xmlspec prose

### ISSUE THREAD

**Issue by michaelhkay** [qt4cg/qtspecs#1336]
In the F+O spec, the `uri-structure-record` appearing in section 6.6, and various other similar record descriptions, are defined using the `fos` namespace markup in xpath-functions.xml. Normally the `fos` XML vocabulary is confined to function-catalog.xml, and is converted to the usual xmlspec vocabulary by the `merge-function-specs` stylesheet.

Using this vocabulary directly within an xmlspec document means that the document doesn't validate against its DTD, and that the `fos` islands aren't validated against the fos.xsd schema.

A better approach here might be to use XInclude to insert text from a separate, schema-validated, document.

**Comment by michaelhkay** [comment]
This is looking pretty messy.

We have three overlapping markup conventions:

**fos:options**

Used for describing the structure of options parameters. 18 occurrences. Always appears in the function catalog within fos:rules, generally as an immediate child but sometimes more deeply nested. The permitted content is defined in fos.xsd. When the element appears as an immediate child of fos:rules, it is validated by virtue of the declaration:

```
<xs:element name="rules">
    <xs:complexType>
      <xs:choice maxOccurs="unbounded">
        <xs:any namespace="##local" processContents="skip"/>
        <xs:element ref="fos:options"/>
        <xs:element ref="fos:record-description"/>
      </xs:choice>
      <xs:attributeGroup ref="fos:diff-markup"/>
    </xs:complexType>
  </xs:element>
```
but when it appears within narrative XHTML it is covered by the `processContents="skip"` wildcard and is therefore not validated.

**fos:record-description**

2 occurrences within function-catalog.xml, both as children of fos:rules. Used for describing records other than options parameters. Structure is defined in fos:xsd, with a very similar but not identical content model to fos:options. Also has three occurrences within xpath-functions.xml (where it is not validated, and not valid). One of these (for parse-html) describes an options parameter.

**fos:record**

Always appears in function-catalog.xml as a child of fos:type, which is referenced from function signatures. 3 occurrences.

**Comment by michaelhkay** [comment]
I suggest that we combine all three elements, using the structure of fos:options as a basis because it is the one that is used most frequently.

I suggest that we use the name `fos:record-description` and make it always appear at the top level of function-catalog.xml (at the same level as fos:function), with a globally-unique name. The fos:option child should be renamed fos:field.

The fos:options element is replaced with a reference to a global fos:record-description.

The fos:record element is also replaced with a reference to a global fos:record-description.

Where fos:record-description appears within xpath-functions.xml, it is replaced with a processing instruction `<?record-description name?>` Note that the stylesheet `merge-functions-specs.xsl` has a template rule with `match="processing-instruction('type')"` that appears to be designed for this purpose, but is currently unused.

**Comment by michaelhkay** [comment]
This is proving quite a hard nut to crack, and unfortunately my attempts to resolve issue #1461 without tackling the bigger problem have made a bit of a mess, leading to builds failing due to intermediate documents being invalid. I think we need to stand back and decide where we want to get to.

1. We have named record types appearing in function signatures. As far as I'm aware the ones we currently use are `uri-structure-record`, `parsed-csv-structure-record`, `load-xquery-module-record`, and `random-number-generator-record`. These need to have a definition somewhere, both so that the reader knows what the record structure is, and so that we can use it semantically, for example when generating test cases. There is an ID/IDREF relationship that needs to be satisfied between the type names and type definitions, both within the function catalog (XSD-defined, fos.xsd), and in the generated file `build/expanded/xpath-functions-40/xpath-functions-40.xml` (DTD-defined).
2. We have within the function catalog `fos:type` entries containing definitions of record types. These appear at the top level (siblings of `fos:function`). There are currently entries for `uri-structure-record`, `parse-html-options`, `parsed-csv-structure-record`. These definitions are incomplete, and are labelled as existing only to satisfy the IDREF constraint.
3. The fos.xsd schema also allows a `fos:options` element. This was designed for defining the content of options parameters, and is widely used for that purpose. The schema explicitly allows it to appear as a child of `fos:rules`, in which case it will be schema-validated; but since fos:rules also allows wildcard `processContents="skip"` content, it can and does appear in other places, in which case it will be accepted without validation. The `fos:options` elements have no name or ID and do not participate in ID/IDREF constraints. The content model overlaps with that of `fos:type` in that both containing field names and types, but in other respects they differ.
4. A `fos:record-description` element has been added more recently. Like `fos:options` this contains a set of `fos:option` elements, and the content models are very similar but not identical. As with `fos:options` it can appear directly as a child of `fos:rules`, or anywhere in which case it is unvalidated. It has a mandatory `id` attribute. Within the function catalog there are currently two instances: `load-query-module-record`, `random-number-generator-record`. Further instances appear directly in xpath-functions.xml, making this file invalid against its DTD: `uri-structure-record`, `parse-html-options`.

**Comment by michaelhkay** [comment]
To keep the amount of change within limits, I propose that we move towards having two markup structures:

1. Options arguments, handled much as at present (and in 3.1). The main change here is that parse-html should use this markup rather than using a fos:record-description. We can make bigger changes here later, but we don't need to do so to solve the current problems.
2. Named record types. This should be a convergence of fos:type and fos:record-description, capturing both the formal structure of the type declaration (as with fos:type) and the associated narrative prose (as with fos:record-description). The definition of the type should be a top-level entry in the function catalog (a sibling of `fos:function`). The formal definition of the type should be expanded below the signature of every function that references the type (duplicated if the same type is used in more than one function). A prose description of the type should be triggered by a `<?record-type NAME?>` processing instruction which may appear anywhere either within xpath-functions.xml or within function-catalog.xml.

**Comment by ndw** [comment]
Thanks for digging into this. That sounds reasonable to me. We can refine in the future, if necessary.

**Comment by michaelhkay** [comment]
Fixed by PR #1465
